packages/firestore/src/local/tab_notification_channel.ts

```diff
@@ -0,0 +1,393 @@
+      'InstanceState.fromJSON',
+      'number',
+      'minMutationBatch',
+      instanceData.minMutationBatch
+    );
+    const maxMutationBatch: number | undefined = validateNamedOptionalType(
+      'InstanceState.fromJSON',
+      'number',
+      'maxMutationBatch',
+      instanceData.minMutationBatch
+    );
+
+    return new InstanceState(
+      instanceId,
+      updateTime,
+      activeTargets,
+      minMutationBatch,
+      maxMutationBatch
+    );
+  }
+}
+
+/**
+ * Instance state of the local instance. Keeps track of all pending mutations.
+ */
+class LocalInstanceState extends InstanceState {
+  private pendingBatches = new SortedSet<BatchId>(primitiveComparator);
+
+  constructor(instanceId: InstanceKey) {
+    super(instanceId, new Date(), new Set<TargetId>(), null, null);
+  }
+
+  get minMutationBatch(): BatchId | null {
+    return this.pendingBatches.first();
+  }
+
+  get maxMutationBatch(): BatchId | null {
+    return this.pendingBatches.last();
+  }
+
+  addPendingMutation(batchId: BatchId): void {
+    assert(
+      !this.pendingBatches.has(batchId),
+      `Batch with ID '${batchId}' already pending.`
+    );
+    this.pendingBatches = this.pendingBatches.add(batchId);
+  }
+
+  removePendingMutation(batchId: BatchId): void {
+    assert(
+      this.pendingBatches.has(batchId),
+      `Pending Batch ID '${batchId}' not found.`
+    );
+    this.pendingBatches = this.pendingBatches.delete(batchId);
+  }
+
+  addQueryTarget(targetId: TargetId): void {
+    assert(
+      !this.activeTargets.has(targetId),
+      `Target with ID '${targetId}' already active.`
+    );
+    this.activeTargets.add(targetId);
+  }
+
+  removeQueryTarget(targetId: TargetId): void {
+    assert(
+      this.activeTargets.has(targetId),
+      `Active Target ID '${targetId}' not found.`
+    );
+    this.activeTargets.delete(targetId);
+  }
+
+  /** Sets the update time to the current time. */
+  markCurrent(): void {
+    this.updateTime = new Date();
+  }
+
+  toJSON(): string {
+    const activeTargets = [];
+    this.activeTargets.forEach(value => activeTargets.push(value));
+
+    const data: InstanceStateSchema = {
+      instanceKey: this.instanceKey,
+      updateTime: this.updateTime.getTime(),
+      activeTargets: Array.from(this.activeTargets)
+    };
+
+    if (this.minMutationBatch !== null) {
+      data.minMutationBatch = this.minMutationBatch;
+      data.maxMutationBatch = this.maxMutationBatch;
+    }
+
+    return JSON.stringify(data);
+  }
+}
+
+/**
+ * `LocalStorageNotificationChannel` uses LocalStorage as the backing store for
+ * the TabNotificationChannel.
+ */
+export class LocalStorageNotificationChannel implements TabNotificationChannel {
+  private activeInstances = new Map<InstanceKey, InstanceState>();
+  private localStorage: Storage;
+  private started = false;
+  private currentInstanceKey: string;
+
+  private storageListener = event => {
```

Comment:
I added the private helper method to clean this up, but kept the local instance variable to allow for removal. I did not test whether `this.handleStorageEvent.bind(this)` constitutes a named function, which would allow me not to do this. Future me is going to verify this in all browsers.